xc/device.ml

```diff
@@ -2336,19 +2353,25 @@ module Backend = struct
         (* Take the first 'max_emulated_nics' elements from the list. *)
         let nics = Xapi_stdext_std.Listext.List.take Dm_Common.max_emulated_nics nics in
 
-        let nics' =
-          if nics <> [] then
-            List.map (fun (mac, bridge, devid) ->
-                [
-                  "-device"; sprintf "rtl8139,netdev=tapnet%d,mac=%s,addr=%d" devid mac (devid + 4);
-                  "-netdev"; sprintf "tap,id=tapnet%d,ifname=tap%d.%d,script=no,downscript=no" devid domid devid
-                ]
-              )
-              nics
-          else
-            [["-net"; "none"]]
-        in
-        common @ misc @ (List.concat nics')
+        let none = ["-net"; "none"] in
+        let add_nic (fds, argv) (mac, bridge, devid) =
+          let ifname          = sprintf "tap%d.%d" domid devid in
+          let uuid, fd as tap = tap_open ifname in
+          let args =
+            [ "-device"; sprintf "rtl8139,netdev=tapnet%d,mac=%s,addr=%d" devid mac (devid + 4)
+            ; "-netdev"; sprintf "tap,id=tapnet%d,fd=%s" devid uuid
+            ] in
+          (tap::fds, args@argv) in
```

Comment:
Looks ok, but as a suggestion to increase readabilitiy, can't we return a type qemu_args here and then use an operator (+) or similar over qemu_args to add this result to common just below?

Reply:
I believe building of the command line requires a refactoring in general. See 

https://github.com/lindig/forkexecd/commit/37d5a22aab257d0387abc8132a00d6bc623fba9f

I could add some comments to explain what is happening here: for each NIC, we introduce command line arguments and open files. These are collected over all NICs in the `fold_left` and in the end need to be added to the already existing command line arguments. 

I don't see how introducing an infix operator would help.